common/src/main/java/com/microsoft/identity/common/internal/providers/microsoft/azureactivedirectory/AzureActiveDirectoryAuthorizationRequest.java

```diff
@@ -22,34 +22,258 @@
+                URLEncoder.encode(getRedirectUri(), ENCODING_UTF8));
+        requestParameters.put(AuthenticationConstants.OAuth2.STATE, encodeProtocolState());
+        requestParameters.put(AuthenticationConstants.OAuth2.SCOPE, SCOPE_OPENID_VALUE);
+
+        // append device and platform info in the query parameters
+        requestParameters.put(ADAL_ID_PLATFORM, ADAL_ID_PLATFORM_VALUE);
+        requestParameters.put(ADAL_ID_VERSION, URLEncoder.encode(mLibraryVersion, ENCODING_UTF8));
+        requestParameters.put(ADAL_ID_OS_VER, URLEncoder.encode(String.valueOf(Build.VERSION.SDK_INT), ENCODING_UTF8));
+        requestParameters.put(ADAL_ID_DM, URLEncoder.encode(android.os.Build.MODEL, ENCODING_UTF8) );
+
+        if (!StringExtensions.isNullOrBlank(mLoginHint)) {
+            requestParameters.put(LOGIN_HINT, URLEncoder.encode(mLoginHint, ENCODING_UTF8) );
+        }
+
+        if (mCorrelationId != null) {
+            requestParameters.put(CLIENT_REQUEST_ID, URLEncoder.encode(mCorrelationId.toString(), ENCODING_UTF8));
+        }
+
+        // Setting prompt behavior to always will skip the cookies for webview.
+        // It is added to authorization url.
+        if (mPromptBehavior == AADPromptBehavior.Always) {
+            requestParameters.put(QUERY_PROMPT, URLEncoder.encode(QUERY_PROMPT_VALUE, ENCODING_UTF8));
+        } else if (mPromptBehavior == AADPromptBehavior.REFRESH_SESSION) {
+            requestParameters.put(QUERY_PROMPT, URLEncoder.encode(QUERY_PROMPT_REFRESH_SESSION_VALUE, ENCODING_UTF8));
+        }
+
+        if (null != mPKCEChallenge) {
+            requestParameters.put(CODE_CHALLENGE, URLEncoder.encode(mPKCEChallenge.getCodeChallenge(), ENCODING_UTF8));
+            // The method used to encode the code_verifier for the code_challenge parameter.
+            requestParameters.put(CODE_CHALLENGE_METHOD, URLEncoder.encode(mPKCEChallenge.getCodeChallengeMethod(), ENCODING_UTF8));
+        }
+
+        // Reading extra qp supplied by developer. append haschrome=1 if developer does not pass as extra qp
+        if (StringExtensions.isNullOrBlank(mExtraQP)
+                || !mExtraQP.contains(AuthenticationConstants.OAuth2.HAS_CHROME)) {
+            requestParameters.put(AuthenticationConstants.OAuth2.HAS_CHROME, "1");
+        }
+
+        // Claims challenge are opaque to the sdk, we're not going to do any merging if both extra qp and claims parameter
+        // contain it. Also, if developer sends it in both places, server will fail it.
+        if (!StringExtensions.isNullOrBlank(mClaimsChallenge)) {
+            requestParameters.put(AuthenticationConstants.OAuth2.CLAIMS, mClaimsChallenge);
+        }
+        if (!StringExtensions.isNullOrBlank(mExtraQP)) {
+            appendExtraQueryParameters(mExtraQP, requestParameters);
+        }
+
+        return requestParameters;
+    }
+
+    private String encodeProtocolState() throws UnsupportedEncodingException {
```

Comment:
Not super clear to me what this function is doing or why -- can you javadoc, rename, add comments, or some combination of those?